client/db/src/changes_tries_storage.rs

```diff
@@ -0,0 +1,1010 @@
+}
+
+/// Persistent struct that contains all the changes tries metadata.
+#[derive(Decode, Encode, Debug)]
+struct ChangesTriesMeta<Block: BlockT> {
+	/// Oldest unpruned max-level (or skewed) digest trie blocks range.
+	/// The range is inclusive from both sides.
+	/// Is None only if:
+	/// 1) we haven't yet finalized any blocks (except genesis)
+	/// 2) if best_finalized_block - min_blocks_to_keep points to the range where changes tries are disabled
+	/// 3) changes tries pruning is disabled
+	pub oldest_digest_range: Option<(NumberFor<Block>, NumberFor<Block>)>,
+	/// End block (inclusive) of oldest pruned max-level (or skewed) digest trie blocks range.
+	/// It is guaranteed that we have no any changes tries before (and including) this block.
+	/// It is guaranteed that all existing changes tries after this block are not yet pruned (if created).
+	pub oldest_pruned_digest_range_end: NumberFor<Block>,
+}
+
+impl<Block: BlockT<Hash=H256>> DbChangesTrieStorage<Block> {
+	/// Create new changes trie storage.
+	pub fn new(
+		db: Arc<dyn KeyValueDB>,
+		meta_column: Option<u32>,
+		changes_tries_column: Option<u32>,
+		key_lookup_column: Option<u32>,
+		header_column: Option<u32>,
+		cache_column: Option<u32>,
+		meta: Arc<RwLock<Meta<NumberFor<Block>, Block::Hash>>>,
+		min_blocks_to_keep: Option<u32>,
+	) -> ClientResult<Self> {
+		let (finalized_hash, finalized_number, genesis_hash) = {
+			let meta = meta.read();
+			(meta.finalized_hash, meta.finalized_number, meta.genesis_hash)
+		};
+		let tries_meta = read_tries_meta(&*db, meta_column)?;
+		Ok(Self {
+			db: db.clone(),
+			meta_column,
+			changes_tries_column,
+			key_lookup_column,
+			header_column,
+			meta,
+			min_blocks_to_keep,
+			cache: DbCacheSync(RwLock::new(DbCache::new(
+				db.clone(),
+				key_lookup_column,
+				header_column,
+				cache_column,
+				genesis_hash,
+				ComplexBlockId::new(finalized_hash, finalized_number),
+			))),
+			build_cache: RwLock::new(ChangesTrieBuildCache::new()),
+			tries_meta: RwLock::new(tries_meta),
+		})
+	}
+
+	/// Commit new changes trie.
+	pub fn commit(
+		&self,
+		tx: &mut DBTransaction,
+		mut changes_trie: MemoryDB<Blake2Hasher>,
+		parent_block: ComplexBlockId<Block>,
+		block: ComplexBlockId<Block>,
+		new_header: &Block::Header,
+		finalized: bool,
+		new_configuration: Option<Option<ChangesTrieConfiguration>>,
+		cache_tx: Option<DbChangesTrieStorageTransaction<Block>>,
+	) -> ClientResult<DbChangesTrieStorageTransaction<Block>> {
+		// insert changes trie, associated with block, into DB
+		for (key, (val, _)) in changes_trie.drain() {
+			tx.put(self.changes_tries_column, &key[..], &val);
+		}
+
+		// if configuration has not been changed AND block is not finalized => nothing to do here
+		let new_configuration = match new_configuration {
+			Some(new_configuration) => new_configuration,
+			None if !finalized => return Ok(DbCacheTransactionOps::empty().into()),
+			None => return self.finalize(
```

Comment:
This is correct, though it may seems a bit strange. This code organization is influenced by how block import operation is organized. The same operation could: (1) just insert block (2) finalize block (3) insert and finalize block (4) insert and finalize block and some its ancestors. This function (commit) is called by any type of import operation => that's why it:
1) does nothing if there's no configuration update && no blocks are finalized;
2) prunes tries (calls finalize) when block is finalized;
3) inserts new configuration into storage when it is updated (if changing block is finalized, `finalize()` is called separately, because it is called for every finalized block)

I do not think there's a better way to reorg this code, unless block import operation will also be reorganized.